File: HotelArtystowApi/Controllers/ZjebleController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MySqlConnector;
using SixLabors.ImageSharp;

using HotelArtystowApi.Models.Entity;
using HotelArtystowApi.Models.Repository;
using HotelArtystowApi.Util.Games;

namespace HotelArtystowApi.Controllers;

[ApiController]
[Route("zjeble")]
public class ZjebleController : ControllerBase
{
    private readonly MySqlDataSource _mysql;

    public ZjebleController([FromServices] MySqlDataSource mysql)
    {
        _mysql = mysql; 
    }

    [HttpGet("getUserSession")]
    [Authorize]
    public async Task<ActionResult<ZjebleUserSessionClearDTO>> GetUserSession()
    {
        int userId = (int)HttpContext.Session.GetInt32("userId")!;

        ZjebleUserSessionRepository repository = new ZjebleUserSessionRepository(_mysql);
        ZjebleRoundRepository roundRepository = new ZjebleRoundRepository(_mysql);

        ZjebleRound round = await roundRepository.GetLatest();
        ZjebleUserSession? session = (await repository.GetBy(new Dictionary<String, dynamic?>() {{"userId", userId}, {"round", round.Id}})).FirstOrDefault();

        if(session is not null)
        {
            session.Round.Value = round;
        }
        else
        {
            session = new ZjebleUserSession();
            session.StartedAt = DateTime.Now;
            session.LivesLeft = 3;
            session.UserId = userId;
            session.Round = new Util.Database.Relation<ZjebleRound>(round.Id, round);

            await repository.Create(session);
        }

        return Ok(session.ToDTO<ZjebleUserSessionClearDTO>());
    }

    [HttpPost("submitAnswer")]
    public async Task<ActionResult> SubmitAnswer([FromBody] String answer) 
    {
        ZjebleRoundRepository roundRepository = new ZjebleRoundRepository(_mysql);
        ZjebleRound round = await roundRepository.GetLatest();

        int userId = (int)HttpContext.Session.GetInt32("userId")!;

        ZjebleUserSessionRepository sessionRepository = new ZjebleUserSessionRepository(_mysql);
        ZjebleUserSession? session = (await sessionRepository.GetBy(new Dictionary<String, dynamic?>() {{"userId", userId}, {"round", round.Id}}))!.FirstOrDefault();

        Dictionary<String, dynamic> response;

        if(session is null)
        {
            response = new Dictionary<String, dynamic>() {
                {"status", false},
                {"message", "Nie znaleziono wygenerowanej sesji"},
                {"noAction", 1}
            };
            return NotFound(response);
        }

        if(session.EndedAt is not null && session.LivesLeft <= 0)
        {
            response = new Dictionary<String, dynamic>() {
                {"status", false},
                {"message", "Nie możesz już zgadywać, wyczerpałeś wszystkie życia ;("},
                {"noAction", 1}
            };
            return Ok(response);
        }
        else if(session.EndedAt is not null)
        {
            response = new Dictionary<String, dynamic>() {
                {"status", false},
                {"message", "Już zgadłeś, czego tu jeszcze szukasz?"},
                {"noAction", 1}
            };
            return Ok(response);
        }

        if(answer != round.Answer)
        {
            session.LivesLeft--;

            if(session.LivesLeft == 0)
                session.EndedAt = DateTime.Now;

            response = new Dictionary<String, dynamic>() {
                {"status", false},
                {"message", "Próbuj dalej"}
            };
        }
        else
        {
            session.EndedAt = DateTime.Now;
            response = new Dictionary<String, dynamic>() {
                {"status", true},
                {"message", "Gratulacje użytkowniku"},
            };

            UserStatisticsRepository statisticsRepository = new UserStatisticsRepository(_mysql);
            UserStatistics stats = (await statisticsRepository.GetBy("userId", userId))!;
            stats.Bees += session.LivesLeft * 10;
            await statisticsRepository.Update(stats);
        }

        await sessionRepository.Update(session);

        return Ok(response);
    }

    [HttpGet("getImageForUser")]
    [Authorize]
    public async Task<ActionResult> getImageForUser()
    {
        int userId = (int)HttpContext.Session.GetInt32("userId")!;

        ZjebleRoundRepository roundRepository = new ZjebleRoundRepository(_mysql);
        ZjebleRound round = await roundRepository.GetLatest();
        ZjebleUserSessionRepository sessionRepository = new ZjebleUserSessionRepository(_mysql);

        Dictionary<String, dynamic?> searchQuery = new Dictionary<String, dynamic?>() {
            {"userId", userId},
            {"round", round.Id}
        };

        ZjebleUserSession session = (await sessionRepository.GetBy(searchQuery))!.First();

        Zjeble zjeble = new Zjeble();

        Image image = (session.EndedAt is null) ? await zjeble.BlurImageAsync(round.PicturePath!, session.LivesLeft * 5) : await zjeble.GetClearImage(round.PicturePath!);
        return File(zjeble.ImageToWebpStream(image), "image/webp");
    }

    [HttpPost("createTodaysRound")]
    public async Task<ActionResult<ZjebleRound>> CreateTodaysRound()
    {
        ZjebleRoundRepository roundRepository = new ZjebleRoundRepository(_mysql);

        ZjebleRound round = await roundRepository.GetLatest();

        if(round.createdAt.DayOfYear == DateTime.Now.DayOfYear)
        {
            return StatusCode(418, "Todays round already exists");
        }

        Zjeble zjeble = new Zjeble();
        String path = zjeble.GetNextImagePath(round);

        round = new ZjebleRound();
        round.Answer = path.Substring(0, path.LastIndexOf('_')).ToLower();
        round.PicturePath = path;
        round.createdAt = DateTime.Now;

        await roundRepository.Create(round);

        return Ok(round);
    }
}

File: api.hotelartystow3.pl/src/Controller/ZjebleController.php
<?php

namespace App\Controller;

use App\Entity\User;
use App\Entity\ZjebleRound;
use App\Entity\ZjebleUserSession;
use App\Repository\ZjebleRoundRepository;
use App\Repository\ZjebleUserSessionRepository;
use App\Service\JsonSerializer;
use App\Service\Zjeble;
use DateTime;
use DateTimeImmutable;
use Doctrine\ORM\EntityManagerInterface;
use Nelmio\ApiDocBundle\Annotation\Model;
use OpenApi\Attributes\Tag;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\Routing\Attribute\Route;
use Symfony\Component\Security\Http\Attribute\CurrentUser;
use OpenApi\Attributes as OA;
use OpenApi\Attributes\JsonContent;
use Symfony\Component\HttpFoundation\BinaryFileResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;

#[Route('/zjeble')]
#[Tag(name: 'Zjeble')]
class ZjebleController extends AbstractController
{
    public function __construct(
        private readonly EntityManagerInterface $em,
        private readonly ZjebleUserSessionRepository $sessionRepository,
        private readonly ZjebleRoundRepository $roundRepository,
        private readonly JsonSerializer $serializer
    ) {}

    #[Route('/getUserSession', name: 'api_zjeble_get_user_session', methods: ['GET'])]
    #[OA\Response(
        response: 200,
        description: 'Logged user session for zjeble',
        content: new Model(type: ZjebleUserSession::class, groups: ['session'])
    )]
    public function getUserSession(#[CurrentUser] User $user)
    {
        $round = $this->roundRepository->getCurrent();

        $session = $this->sessionRepository->getForRound($user, $round);

        if(empty($session))
        {
            $session = new ZjebleUserSession();
            $session
                ->setStartedAt(new \DateTimeImmutable())
                ->setLivesLeft(3)
                ->setUser($user)
                ->setRound($round);

            try
            {
                $this->em->beginTransaction();
                $this->em->persist($session);
                $this->em->flush();
                $this->em->commit();
            }
            catch(\Exception $e)
            {
                $this->em->rollback();

                return new JsonResponse(['message' => 'Nie udało się pobrać sesji na dzisiaj'], 500);
            }
        }

        return $this->serializer->res($session, ['session']);
    }

    #[Route('/submitAnswer', name: 'api_zjeble_submit_answer', methods: ['POST'])]
    #[OA\RequestBody(
        required: true,
        content: new JsonContent(
            example: ['answer' => 'string']
        )
    )]
    #[OA\Response(
        response: 200,
        description: 'Submit answer for zjeble',
        content: new JsonContent(
            example: [
                'status' => 'true|false',
                'message' => 'string',
                'noAction' => '?int'
            ]
        )
    )]
    public function submitAnswer(#[CurrentUser] User $user, Request $request)
    {
        [
            'answer' => $answer
        ] = $request->toArray();

        $round = $this->roundRepository->getCurrent();
        $session = $this->sessionRepository->getForRound($user, $round);

        if(empty($session))
            return new JsonResponse(['status' => false, 'message' => 'Nie znaleziono wygenerowanej sesji', 'noAction' => 1], 404);

        if(!empty($session->getEndedAt()) && $session->getLivesLeft() <= 0)
            return new JsonResponse(['status' => false, 'message' => 'Nie możesz już zgadywać, wyczerpałeś wszystkie życia ;(', 'noAction' => 1]);

        if(!empty($session->getEndedAt()))
            return new JsonResponse(['status' => false, 'message' => 'Już zgadłeś, czego tu jeszcze szukasz?', 'noAction' => 1]);

        $lives = $session->getLivesLeft();

        $response = [];
        if($answer !== $round->getAnswer())
        {
            $lives--;

            if($lives === 0)
                $session->setEndedAt(new \DateTimeImmutable());

            $session->setLivesLeft($lives);

            $response = [
                'status' => false,
                'message' => 'Próbuj dalej'
            ];
        }
        else
        {
            $session->setEndedAt(new \DateTimeImmutable());
            $response = [
                'status' => true,
                'message' => 'Gratulacje użytkowniku'
            ];

            $statistics = $user->getUserStatistics();
            $beesEarned = $session->getLivesLeft() * 10;

            $statistics->setBees($statistics->getBees() + $beesEarned);
        }

        try
        {
            $this->em->beginTransaction();
            $this->em->persist($session);

            if(isset($statistics))
                $this->em->persist($statistics);

            $this->em->flush();
            $this->em->commit();
        }
        catch(\Exception $e)
        {
            $this->em->rollback();
            return new JsonResponse(['false' => 'Nie udało się zapisać w bazie', 'noAction' => 1]);
        }

        return new JsonResponse($response);
    }

    #[Route('/getImageForUser', name: 'api_zjeble_get_image_for_user', methods: ['GET'])]
    public function getImageForUser(#[CurrentUser] User $user, Zjeble $zjeble)
    {
        $round = $this->roundRepository->getCurrent();
        $session = $this->sessionRepository->getForRound($user, $round);

        $image = empty($session->getEndedAt()) ? $zjeble->blurImage($round->getPicturePath(), $session->getLivesLeft() * 3) : $zjeble->getClearImage($round->getPicturePath());

        return new Response($image, 200, [
            'Content-Type' => 'image/webp'
        ]);
    }

    #[Route('/createTodaysRound', name: 'api_zjeble_create_todays_round', methods: ['GET'])]
    public function createTodaysRound(Zjeble $zjeble)
    {
        $round = null;
        try
        {
            $round = $this->roundRepository->getCurrent();
        }
        catch(\Exception $e) {
            $round = new ZjebleRound();
            $round->setPhotoIndex(0)->setCreatedAt(new \DateTimeImmutable('-1 day'));
        }

        if($round->getCreatedAt()->format('z') === (new \DateTime('+10 minutes'))->format('z'))
            return new JsonResponse(['status' => false, 'message' => 'Todays round already exists'], 418);

        try
        {
            $res = $zjeble->createNextRound($round);
            return new JsonResponse(['status' => true], 200);
        }
        catch(\Exception $e)
        {
            return new JsonResponse(['status' => false, 'message' => $e->getMessage()], 500);
        }
    }
}
